read-fonts/src/tables/colr.rs

```diff
@@ -1,5 +1,696 @@
+        // Note: FreeType never returns an error for these lookups, so
+        // we do the same and just `unwrap_or_default` on var store
+        // errors.
+        // See <https://gitlab.freedesktop.org/freetype/freetype/-/blob/fc01e7dd/src/sfnt/ttcolr.c#L574>
+        let mut deltas = [0; N];
+        if self.coords.is_empty()
+            || self.var_store.is_none()
+            || var_index_base == NO_VARIATION_DELTAS
+        {
+            return deltas;
+        }
+        let var_store = self.var_store.as_ref().unwrap();
+        if let Some(index_map) = self.index_map.as_ref() {
+            for (i, delta) in deltas.iter_mut().enumerate() {
+                let var_index = var_index_base + i as u32;
+                *delta = index_map
+                    .get(var_index)
+                    .and_then(|delta_index| var_store.compute_delta(delta_index, self.coords))
+                    .unwrap_or_default();
+            }
+        } else {
+            for (i, delta) in deltas.iter_mut().enumerate() {
+                let var_index = var_index_base + i as u32;
+                // If we don't have a var index map, use our index as the inner
+                // component and set the outer to 0.
+                let delta_index = DeltaSetIndex {
+                    outer: 0,
+                    inner: var_index as u16,
+                };
+                *delta = var_store
+                    .compute_delta(delta_index, self.coords)
+                    .unwrap_or_default();
+            }
+        }
+        deltas
+    }
+}
+
+impl<'a> Deref for ColrInstance<'a> {
+    type Target = Colr<'a>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.colr
+    }
+}
+
+impl<'a> ClipBox<'a> {
+    /// Resolves a clip box, applying variation deltas using the given
+    /// instance.
+    pub fn resolve(&self, instance: &ColrInstance<'a>) -> BoundingBox<Fixed> {
+        match self {
+            Self::Format1(cbox) => BoundingBox {
+                x_min: cbox.x_min().to_fixed(),
+                y_min: cbox.y_min().to_fixed(),
+                x_max: cbox.x_max().to_fixed(),
+                y_max: cbox.y_max().to_fixed(),
+            },
+            Self::Format2(cbox) => {
+                let deltas = instance.var_deltas::<4>(cbox.var_index_base());
+                BoundingBox {
+                    x_min: cbox.x_min().apply_delta(deltas[0]),
+                    y_min: cbox.y_min().apply_delta(deltas[1]),
+                    x_max: cbox.x_max().apply_delta(deltas[2]),
+                    y_max: cbox.y_max().apply_delta(deltas[3]),
+                }
+            }
+        }
+    }
+}
+
+/// Simplified version of a [`ColorStop`] or [`VarColorStop`] with applied
+/// variation deltas.
+#[derive(Clone, Debug)]
+pub struct ResolvedColorStop {
+    pub offset: Fixed,
+    pub palette_index: u16,
+    pub alpha: Fixed,
+}
+
+/// Collection of [`ColorStop`] or [`VarColorStop`].
+pub struct ColorStops<'a> {
```

Comment:
if it _either_ contains one _or_ the other, never both, should this be a enum with two variants?

Reply:
no problems, I just asked out of curiousity (and based on my understanding that color stops can only ever be defined as either all variable or none at all). Just do whatever feels more convenient here